src/riscv/parser.rs

```diff
@@ -98,33 +101,50 @@ pub fn parse_asm(input: &str) -> Vec<Statement> {
         .collect()
 }
 
-pub fn extract_labels(statements: &[Statement]) -> BTreeSet<&str> {
+pub fn extract_label_offsets(statements: &[Statement]) -> BTreeMap<&str, usize> {
     statements
         .iter()
-        .filter_map(|s| match s {
-            Statement::Label(l) => Some(l.as_str()),
+        .enumerate()
+        .filter_map(|(i, s)| match s {
+            Statement::Label(l) => Some((l.as_str(), i)),
             Statement::Directive(_, _) | Statement::Instruction(_, _) => None,
         })
         .collect()
 }
 
-pub fn extract_label_references(statements: &[Statement]) -> BTreeSet<&str> {
-    statements
-        .iter()
-        .flat_map(|s| match s {
-            Statement::Label(_) | Statement::Directive(_, _) => None,
-            Statement::Instruction(_, args) => Some(args.iter().filter_map(|arg| match arg {
+pub fn referenced_labels(statement: &Statement) -> BTreeSet<&str> {
+    match statement {
+        Statement::Label(_) | Statement::Directive(_, _) => Default::default(),
+        Statement::Instruction(_, args) => args
+            .iter()
+            .filter_map(|arg| match arg {
                 Argument::Register(_) | Argument::StringLiteral(_) => None,
                 Argument::Symbol(s) => Some(s.as_str()),
                 Argument::RegOffset(_, c) | Argument::Constant(c) => match c {
                     Constant::Number(_) => None,
                     Constant::HiDataRef(s) | Constant::LoDataRef(s) => Some(s.as_str()),
                 },
                 Argument::Difference(_, _) => todo!(),
-            })),
-        })
-        .flatten()
-        .collect()
+            })
+            .collect(),
+    }
+}
+
+pub fn ends_control_flow(s: &Statement) -> bool {
+    match s {
+        Statement::Instruction(instruction, _) => match instruction.as_str() {
+            "li" | "lui" | "mv" | "add" | "addi" | "sub" | "neg" | "mul" | "mulhu" | "xor"
+            | "xori" | "and" | "andi" | "or" | "ori" | "not" | "slli" | "sll" | "srli" | "srl"
+            | "seqz" | "snez" | "slti" | "sltu" | "sltiu" | "beq" | "beqz" | "bgeu" | "bltu"
+            | "blt" | "bge" | "bltz" | "blez" | "bgtz" | "bgez" | "bne" | "bnez" | "jal"
+            | "jalr" | "call" | "ecall" | "lw" | "lb" | "lbu" | "sw" | "sh" | "sb" => false,
+            "j" | "jr" | "tail" | "ret" | "unimp" => true,
```

Comment:
Why are `jal` and `jalr` not control-flow-ending?

Reply:
It's a `call`, it stores the address of the following instruction in a register.

Reply:
`jal` is "jump and link"